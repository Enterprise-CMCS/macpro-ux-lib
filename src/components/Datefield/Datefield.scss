@use "../../assets/theme/macbis-colors" as c;
@import url("https://fonts.googleapis.com/css?family=Open+Sans");

.datefield {
  .usa-hint,
  .usa-label {
    font-family: "Open Sans", sans-serif;
  }

  .usa-hint.input-error {
    color: c.$error;
  }

  .usa-date-picker {
    .calendar-button {
      background-color: transparent;
      border: 0;
    }

    .calendar-button:enabled {
      cursor: pointer;

      :hover {
        background-color: c.$gray-lightest;
      }
    }

    .react-calendar {
      position: absolute;
      font-family: "Open Sans", sans-serif;
      background-color: c.$gray-lightest;
      left: 188px;

      .react-calendar__navigation__arrow {
        font-size: 30px;
      }

      .react-calendar__month-view__weekdays__weekday > abbr {
        font-size: 16px;
        text-decoration: none;
      }

      .react-calendar__tile:not(.react-calendar__tile--active, :disabled, .range-start-end) {
        color: black;
      }

      .react-calendar__month-view__days__day--weekend,
      .react-calendar__month-view__days__day--neighboringMonth {
        color: rgba(16, 16, 16, 0.3);
      }

      .react-calendar__tile--active.react-calendar__month-view__days__day--weekend:not(:disabled) {
        color: white;
      }
    }

    .react-calendar__tile--now:not(.react-calendar__tile--active, .range-start-end, .range-selected),
    .grey-lightest {
      background-color: c.$gray-lightest;
    }
  }

  .range-selected,
  .range-selected:hover {
    background-color: c.$primary-lighter !important;
  }
  .range-start-end,
  .range-start-end:hover {
    background-color: c.$primary !important;
    color: white;
  }
}
