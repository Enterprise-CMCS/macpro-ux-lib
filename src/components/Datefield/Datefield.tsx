import React, { useLayoutEffect, useRef, useState } from "react";
import {
  completeDateFilter,
  numbersAndSlashesFilter,
  checkValidMonthDays,
  splitDateIntoVariables,
  formatPropDates,
} from "../../utils";

import datePicker from "../../../node_modules/@uswds/uswds/packages/usa-date-picker/src";

type IntrinsicElements = JSX.IntrinsicElements["input"];
export interface Props extends IntrinsicElements {
  id: string;
  name: string;
  label: string;
  hint?: boolean;
  disabled?: boolean;
  minDate?: string;
  maxDate?: string;
  value?: string;
  required?: boolean;
}

/**
 * Datefield Component
 * @param {string}  id                     A unique identifier for the input.
 * @param {string}  name                   Name of the input field.
 * @param {string}  label                  Field label.
 * @param {boolean} [hint]                 Boolean that shows or hide the date format hint, in the format mm/dd/yyyy.
 * @param {string}  [disabled]             Controls whether or not the date picker is disabled to the user.
 * @param {string}  [minDate]              The date picker will not allow a date selection before this date. The date should be in the format mm/dd/yyyy
 * @param {string}  [maxDate]              The date picker will not allow a date selection after this date. The date should be in the format mm/dd/yyyy.
 * @param {string}  [value]                Value of the input element.
 * @param {string}  [required]             The date picker component will be required in terms of native form validation.
 */

export const Datefield: React.FC<Props> = ({
  id,
  name,
  label,
  value,
  minDate,
  maxDate,
  hint = true,
  disabled = false,
  required = false,
  ...rest
}) => {
  value = completeDateFilter.test(value || "") ? value : "";
  const [dateError, setDateError] = useState(false);
  const datePickerRef = useRef<HTMLDivElement>(null);

  useLayoutEffect(() => {
    const datePickerElement = datePickerRef.current;
    if (typeof datePicker.on === "function") {
      datePicker.on(datePickerElement);

      const dateFieldinput = Array.from(
        document.getElementsByClassName("usa-date-picker__external-input")
      ).find((dateField) => dateField.id === id);

      if (dateFieldinput) {
        dateFieldinput.addEventListener("keydown", (e: any) => filterInput(e));
        dateFieldinput.addEventListener("blur", (e: any) =>
          checkValidDate(e.target.value)
        );
      }
    }

    return () => {
      if (typeof datePicker.off === "function") {
        datePicker.off(datePickerElement);
      }
    };
  }, [value]);

  const filterInput = (typedValue: KeyboardEvent) => {
    if (
      (typedValue && numbersAndSlashesFilter.test(typedValue.key)) ||
      typedValue.key === "Backspace" ||
      typedValue.key === "Tab" ||
      typedValue.metaKey
    ) {
      setDateError(false);
    } else {
      typedValue.preventDefault();
    }
  };

  const checkValidDate = (date: string) => {
    let [month, day, year] = splitDateIntoVariables(date);

    if (
      (completeDateFilter.test(date) &&
        checkValidMonthDays(parseInt(month), parseInt(year), parseInt(day))) ||
      date === ""
    ) {
      setDateError(false);
    } else {
      setDateError(true);
    }
  };

  return (
    <div className="usa-form-group datefield" ref={datePickerRef}>
      <label className="usa-label" id={`${id}-label`} htmlFor={id}>
        {label}
      </label>
      {(hint || dateError) && (
        <div
          className={`usa-hint${dateError ? " input-error" : ""}`}
          id={`${id}-hint`}
        >
          {`${
            dateError ? "Inputted date must be a valid date in " : ""
          }mm/dd/yyyy`}
        </div>
      )}

      <div
        className="usa-date-picker"
        data-min-date={minDate !== undefined && formatPropDates(minDate)}
        data-max-date={maxDate !== undefined && formatPropDates(maxDate)}
        data-default-value={formatPropDates(value)}
      >
        <input
          className="usa-input"
          id={id}
          name={name}
          aria-labelledby={`${id}-label`}
          aria-describedby={hint ? `${id}-hint` : `${id}-label`}
          disabled={disabled}
          {...rest}
        />
      </div>
    </div>
  );
};
