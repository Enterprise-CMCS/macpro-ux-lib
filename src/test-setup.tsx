import React, { ReactElement } from "react";
import { render, RenderOptions } from "@testing-library/react";
import "@testing-library/jest-dom";

// Here we can set all providers for themes, stores and other stuff
const AllTheProviders = ({ children }: { children: JSX.Element }) => {
  return <>{children}</>;
};

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, "wrapper">
) => render(ui, { wrapper: AllTheProviders, ...options });

export * from "@testing-library/react";
export { customRender as render };
