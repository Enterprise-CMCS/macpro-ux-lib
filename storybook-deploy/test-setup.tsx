import React, { ReactElement } from "react";
import { render, RenderOptions } from "@testing-library/react";
import "@testing-library/jest-dom";

// Here we can set all providers for themes, stores and other stuff
const AllTheProviders = ({ children }: { children: JSX.Element }) => {
  return <>{children}</>;
};

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, "wrapper">
) => render(ui, { wrapper: AllTheProviders, ...options });

export * from "@testing-library/react";
export { customRender as render };

/**
 * Clean up attributes from the rendered component by removing the generated ids.
 *
 * Before:
 *   for="input-type-text-123456"
 *   id="input-type-text-123456"
 *
 * After:
 *   for="input-type-text"
 *   id="input-type-text"
 *
 * This is useful during snapshot testing. Snapshot comparisons will fail without addressing these items.
 *
 * @param {HTMLElement} container The component to clean up.
 * @param {prefixes}      string[]  The reference substrings to leave alone.
 */
