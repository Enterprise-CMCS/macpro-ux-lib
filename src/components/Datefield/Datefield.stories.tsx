import React from "react";
import { ComponentStory, ComponentMeta } from "@storybook/react";
import { Datefield } from "./Datefield";
import { USWDSDecorator } from "../../utils";

export default {
  title: "COMPONENTS/Datefield",
  component: Datefield,
  argTypes: {
    id: {
      description: "A unique identifier for the input.",
    },
    fieldName: {
      description: "Name of the input field.",
    },
    label: {
      description: "Field label.",
    },
    hint: {
      description:
        "Boolean that shows or hide the date format hint, in the format YYYY-MM-DD.",
    },
    required: {
      description: "Adds semantic required.",
    },
    disabled: {
      description:
        "Controls whether or not the date picker is disabled to the user.",
    },
    defaultDate: {
      control: { type: "date" },
      description:
        "The date picker input will set this value if it is a valid date. The date should be in the format YYYY-MM-DD.",
    },
    minDate: {
      description:
        "The date picker will not allow a date selection before this date. The date should be in the format YYYY-MM-DD.",
    },
    maxDate: {
      description:
        "The date picker will not allow a date selection after this date. The date should be in the format YYYY-MM-DD.",
    },
    value: {
      description: "Value of the input element.",
    },
  },
  args: {},
  parameters: {
    docs: {
      description: {
        component:
          "A date picker helps users select a single date. Use this when users need to schedule or record an event and benefit from the context of a calendar. It can also be used when knowing the day of the week helps users choose a specific date.",
      },
    },
  },
  decorators: [...USWDSDecorator],
} as ComponentMeta<typeof Datefield>;

const Template: ComponentStory<typeof Datefield> = ({ ...rest }) => (
  <Datefield {...rest} />
);

export const PrimaryDatefield = Template.bind({});
PrimaryDatefield.args = {};

export const DefaultDatefield = Template.bind({});
DefaultDatefield.args = { defaultDate: "2022-08-08" };
DefaultDatefield.parameters = {
  docs: {
    description: {
      story:
        "Default place holder values can be giving to inputs that could make selecting dates easier for our users",
    },
  },
};

export const MinimumDateRange = Template.bind({});
MinimumDateRange.args = { minDate: "2022-08-08" };
MinimumDateRange.parameters = {
  docs: {
    description: {
      story:
        "A min date can be provided to prevent users from selecting a date before a certain date",
    },
  },
};

export const MaximumDateRange = Template.bind({});
MaximumDateRange.args = { maxDate: "2022-08-08" };
MaximumDateRange.parameters = {
  docs: {
    description: {
      story:
        "A max date can be provided to prevent users from selecting a date after a certain date",
    },
  },
};
