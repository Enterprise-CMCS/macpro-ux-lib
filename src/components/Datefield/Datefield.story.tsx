import React from "react";
import { ComponentStory, ComponentMeta } from "@storybook/react";
import { Datefield } from "./Datefield";

export default {
  title: "USWDS/Base/Datefield",
  component: Datefield,
  argTypes: {},
  args: {},
} as ComponentMeta<typeof Datefield>;

const Template: ComponentStory<typeof Datefield> = ({ ...rest }) => (
  <Datefield {...rest} />
);

export const PrimaryDatefield = Template.bind({});
PrimaryDatefield.args = {};
