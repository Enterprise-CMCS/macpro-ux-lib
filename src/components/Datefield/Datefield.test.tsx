import React from "react";
import { Datefield } from "./Datefield";
import { fireEvent, screen, render } from "../../test-setup";

describe("Tests for the Datefield component.", () => {
  it("Should render with hint", () => {
    render(
      <Datefield
        data-testid="Datefield"
        id="test-1"
        fieldName="test-1"
        label="test-1"
        hint
      />
    );
    const DatefieldComp = screen.getByTestId("Datefield");

    fireEvent.click(DatefieldComp);
    expect(DatefieldComp).toBeInTheDocument();
    expect(DatefieldComp).toHaveTextContent("mm/dd/yyyy");
  });
});

describe("compontent snapshots", () => {
  it("primary datefield", () => {
    const { container } = render(
      <Datefield hint id="test-3" fieldName="test-3" label="test-3" />
    );
    expect(container).toMatchSnapshot();
  });
});
