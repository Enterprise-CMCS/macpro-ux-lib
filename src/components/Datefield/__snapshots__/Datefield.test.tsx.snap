// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`compontent snapshots primary datefield 1`] = `
<div>
  <div
    class="usa-form-group"
  >
    <label
      class="usa-label"
      for="test-3"
      id="test-3-label"
    >
      test-3
    </label>
    <div
      class="usa-hint"
      id="test-3-hint"
    >
      mm/dd/yyyy
    </div>
    <div
      class="usa-date-picker"
    >
      <input
        aria-describedby="test-3-hint"
        aria-labelledby="test-3-label"
        class="usa-input"
        id="test-3"
        name="test-3"
        value=""
      />
    </div>
  </div>
</div>
`;
